client/matching/partition_config_provider.go

```diff
@@ -0,0 +1,179 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+//go:generate mockgen -package $GOPACKAGE -source $GOFILE -destination partition_config_provider_mock.go -package matching github.com/uber/cadence/client/matching PartitionConfigProvider
+
+package matching
+
+import (
+	"sync"
+
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/types"
+)
+
+type (
+	// PartitionConfigProvider is the interface for implementers of
+	// component that provides partition configuration for task list
+	// partitions
+	PartitionConfigProvider interface {
+		// GetNumberOfReadPartitions returns the number of read partitions
+		GetNumberOfReadPartitions(domainID string, taskList types.TaskList, taskListType int) int
+		// GetNumberOfWritePartitions returns the number of write partitions
+		GetNumberOfWritePartitions(domainID string, taskList types.TaskList, taskListType int) int
+		// UpdatePartitionConfig updates the partition configuration for a task list
+		UpdatePartitionConfig(domainID string, taskList types.TaskList, taskListType int, config *types.TaskListPartitionConfig)
+	}
+
+	syncedTaskListPartitionConfig struct {
+		sync.RWMutex
+		types.TaskListPartitionConfig
+	}
+
+	partitionConfigProviderImpl struct {
+		configCache         cache.Cache
+		logger              log.Logger
+		domainIDToName      func(string) (string, error)
+		enableReadFromCache dynamicconfig.BoolPropertyFnWithTaskListInfoFilters
+		nReadPartitions     dynamicconfig.IntPropertyFnWithTaskListInfoFilters
+		nWritePartitions    dynamicconfig.IntPropertyFnWithTaskListInfoFilters
+	}
+)
+
+func (c *syncedTaskListPartitionConfig) updateConfig(newConfig types.TaskListPartitionConfig) {
+	c.Lock()
+	defer c.Unlock()
+	if c.Version < newConfig.Version {
+		c.TaskListPartitionConfig = newConfig
+	}
+}
+
+func NewPartitionConfigProvider(
+	logger log.Logger,
+	domainIDToName func(string) (string, error),
+	dc *dynamicconfig.Collection,
+) PartitionConfigProvider {
+	return &partitionConfigProviderImpl{
+		logger:              logger,
+		domainIDToName:      domainIDToName,
+		enableReadFromCache: dc.GetBoolPropertyFilteredByTaskListInfo(dynamicconfig.MatchingEnableGetNumberOfPartitionsFromCache),
+		nReadPartitions:     dc.GetIntPropertyFilteredByTaskListInfo(dynamicconfig.MatchingNumTasklistReadPartitions),
+		nWritePartitions:    dc.GetIntPropertyFilteredByTaskListInfo(dynamicconfig.MatchingNumTasklistWritePartitions),
+		configCache: cache.New(&cache.Options{
+			TTL:             0,
```

Comment:
why 0 ttl?

Reply:
It means no ttl.

Reply:
I think some ttl would be nice. Are we ignoring sticky & bits-like temporary tasklist here?

Reply:
We're only tracking normal tasklist. I'm not sure if bits-like tasklist are treated as normal tasklist or dynamic tasklist now.